packages/utils/src/types/filters.ts

```diff
@@ -39,41 +22,175 @@ export type NumericOperator =
+})
+
 export type StringOperator = {
   $regex?: string
 }
 
+export const StringOperatorSchema = z.object({
+  $regex: z.string().optional(),
+})
+
 export type FilterOperator =
   | LogicalOperator
   | NumericOperator
   | ArrayOperator
   | StringOperator
 
+export type ArrayOperator =
+  | {
+      $some?: FilterOperator[]
+    }
+  | {
+      $first?: FilterOperator
+    }
+  | {
+      $last?: FilterOperator
+    }
+  | {
+      $nth?: NthFilter
+    }
+
+export const ArrayOperatorSchema: z.ZodType<ArrayOperator> = z.union([
+  z.object({
+    $some: z
+      .lazy(() => FilterOperatorSchema)
+      .array()
+      .optional(),
+  }),
+  z.object({ $first: z.lazy(() => FilterOperatorSchema).optional() }),
+  z.object({ $last: z.lazy(() => FilterOperatorSchema).optional() }),
+  z.object({ $nth: z.lazy(() => NthFilterSchema).optional() }),
+])
+
+export type LogicalOperator =
+  | {
+      $and?: FilterOperator[]
+    }
+  | {
+      $or?: FilterOperator[]
+    }
+
+export const LogicalOperatorSchema: z.ZodType<LogicalOperator> = z.union([
+  z.object({
+    $and: z
+      .lazy(() => FilterOperatorSchema)
+      .array()
+      .optional(),
+  }),
+  z.object({
+    $or: z
+      .lazy(() => FilterOperatorSchema)
+      .array()
+      .optional(),
+  }),
+])
+
+export const FilterOperatorSchema = z.union([
+  LogicalOperatorSchema,
+  NumericOperatorSchema,
+  ArrayOperatorSchema,
+  StringOperatorSchema,
+])
+
 export type TransactionFilter = {
   [K in keyof Transaction]: FilterOperator
 }
 
+export const TransactionFilterSchema = z.record(
+  z.string(),
+  FilterOperatorSchema,
+)
+
 type Primitive = string | number | boolean
+export const PrimitiveSchema = z.union([z.string(), z.number(), z.boolean()])
+
 export type FilterObject = {
   [key: string]: Filter
 }
-export type BitmaskFilter = {
-  bitmask: bigint | number | string
-  value: bigint | number | string | NumericOperator
-}
-export type NthFilter = {
-  index: bigint | number | string
-  value: TransactionFilter | FilterObject
-}
-export type Filter = Primitive | FilterObject | FilterArray | Abi
-export type FilterArray = Filter[]
+
+export const FilterObjectSchema: z.ZodType<FilterObject> = z.record(
+  z.string(),
+  z.lazy(() => FilterSchema),
+)
+
 export interface AbiFilter extends FilterObject {
   $abi: Abi
 }
+export const AbiFilterSchema = z
+  .object({
+    $abi: AbiSchema,
+  })
+  .catchall(z.lazy(() => FilterSchema))
 
 export interface AbstractAbiFilter extends FilterObject {
   $abiAbstract: Abi
 }
+export const AbstractAbiFilterSchema = z
+  .object({
+    $abiAbstract: AbiSchema,
+  })
+  .catchall(z.lazy(() => FilterSchema))
+
 export interface AbiParamFilter extends FilterObject {
   $abiParams: string[]
 }
+
+export const AbiParamFilterSchema = z
+  .object({
+    $abiParams: AbiSchema,
+  })
+  .catchall(z.lazy(() => FilterSchema))
+
+export type Filter =
+  | Primitive
+  | FilterObject
+  | FilterArray
+  | FilterOperator
```

Comment:
@Quazia is this correct? I added the `FilterOperator` to `Filter` union type, and schema. Unsure if it's the move.

Reply:
I think that is correct - some of the issues before may have been from _not_ having it in this.